Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean

```diff
@@ -0,0 +1,687 @@
+          succ (c.emb i a') ≠ cases (motive := fun _ ↦ Fin (n + 1)) 0 (succ ∘ c.emb j) b' by
+        apply Set.disjoint_iff_forall_ne.2
+        simp only [hij, ↓reduceDIte, mem_range, comp_apply, ne_eq, forall_exists_index,
+          forall_apply_eq_imp_iff]
+        intro a b
+        apply this
+      intro a' b'
+      induction b' using Fin.induction with
+      | zero => simpa using succ_ne_zero (c.emb i a')
+      | succ b' =>
+        simp only [Nat.succ_eq_add_one, cases_succ, comp_apply, ne_eq, succ_inj]
+        apply c.emb_ne_emb_of_ne hij
+    · simp only [onFun, h, ↓reduceDIte, hj]
+      apply Set.disjoint_iff_forall_ne.2
+      simp only [mem_range, comp_apply, ne_eq, forall_exists_index, forall_apply_eq_imp_iff,
+        succ_inj]
+      intro a b
+      apply c.emb_ne_emb_of_ne hij
+  cover := by
+    refine Fin.cases ?_ (fun i ↦ ?_)
+    · simp only [mem_iUnion, mem_range]
+      exact ⟨k, ⟨0, by simp⟩, by simp⟩
+    · simp only [mem_iUnion, mem_range]
+      rcases eq_or_ne (c.index i) k with rfl | hi
+      · have A : update c.partSize (c.index i) (c.partSize (c.index i) + 1) (c.index i) =
+          c.partSize (c.index i) + 1 := by simp
+        exact ⟨c.index i, cast A.symm (succ (c.invEmbedding i)), by simp⟩
+      · have A : update c.partSize k (c.partSize k + 1) (c.index i) = c.partSize (c.index i) := by
+          simp [hi]
+        exact ⟨c.index i, cast A.symm (c.invEmbedding i), by simp [hi]⟩
+
+lemma index_extendMiddle_zero (c : OrderedFinpartition n) (i : Fin c.length) :
+    (c.extendMiddle i).index 0 = i := by
+  have : (c.extendMiddle i).emb i 0 = 0 := by simp [extendMiddle]
+  conv_rhs at this => rw [← (c.extendMiddle i).emb_invEmbedding 0]
+  contrapose! this
+  exact (c.extendMiddle i).emb_ne_emb_of_ne (Ne.symm this)
+
+lemma range_emb_extendMiddle_ne_singleton_zero (c : OrderedFinpartition n) (i j : Fin c.length) :
+    range ((c.extendMiddle i).emb j) ≠ {0} := by
+  intro h
+  rcases eq_or_ne j i with rfl | hij
+  · have : Fin.succ (c.emb j 0) ∈ ({0} : Set (Fin n.succ)) := by
+      rw [← h]
+      simp only [Nat.succ_eq_add_one, mem_range]
+      have A : (c.extendMiddle j).partSize j = c.partSize j + 1 := by simp [extendMiddle]
+      refine ⟨Fin.cast A.symm (succ 0), ?_⟩
+      simp only [extendMiddle, ↓reduceDIte, comp_apply, cast_trans, cast_eq_self, cases_succ]
+    simp only [mem_singleton_iff] at this
+    exact Fin.succ_ne_zero _ this
+  · have : (c.extendMiddle i).emb j 0 ∈ range ((c.extendMiddle i).emb j) :=
+      mem_range_self 0
+    rw [h] at this
+    simp only [extendMiddle, hij, ↓reduceDIte, comp_apply, cast_zero, mem_singleton_iff] at this
+    exact Fin.succ_ne_zero _ this
+
+/-- Extend an ordered partition of `n` entries, by adding singleton to the left or appending it
+to one of the existing part. -/
+def extend (c : OrderedFinpartition n) (i : Option (Fin c.length)) : OrderedFinpartition (n + 1) :=
+  match i with
+  | none => c.extendLeft
+  | some i => c.extendMiddle i
```

Comment:
Given that you are only using this in `extendEquiv`, why not inline it? This is what you do for the `erase` side

Reply:
It's for the way I use it in the sequel. You can have a look at #17400 if you want: `c.extend i` shows up a lot of times.